File: 3020. Find the Maximum Number of Elements in Subset.py
# 3020. Find the Maximum Number of Elements in Subset


from collections import Counter


class Solution:
    def maximumLength(self, nums: List[int]) -> int:
        count = Counter(nums)
        ans = 0
        for n in count:
            if n == 1:
                ans = max(ans, count[n] if count[n] % 2 else count[n] -1)
            else:
                cur = 0
                while n ** 2 in count and count[n] >= 2:
                   cur += 2
                   n = n**2
                cur += 1
                ans = max(cur, ans)
        return ans

File: 3020. Find the Maximum Number of Elements in Subset.cpp
// 3020. Find the Maximum Number of Elements in Subset

#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;


class Solution {
    public:
        int maximumLength(vector<int>& nums) {
            unordered_map<long long, int> dic;
            int ans {0};
    
            for (int n: nums) {
                dic[n]++;
            }
    
            for (auto& [n, freq] : dic) {
                int cur = 0;
                if (n == 1) {
                    cur = dic[n] % 2 ? dic[n] : dic[n] - 1; 
                }
                else {
                    long long num = n;
                    while (dic.count(num*num) && dic[num] >= 2 && num <= 1e9) {
                        cur += 2;
                        num *= num;
                    }
                    cur += 1;
                }
                ans = max(cur, ans);
            }
            return ans;
        }
    };
    